Add tests for organisation schema test cases

dalepotter opened the pull request:
This adds the `iati-organisations-schema.xsd` file to the testrunner.

Code requires refactoring to make more DRY.

dalepotter committed:
Add tests for organisation schema test cases

This adds the `iati-organisations-schema.xsd` file to the testrunner.

Code requires refactoring to make more DRY.

hayfield committed:
Remove non-XML test files

Test data files should be valid XML. Several were not.

Fixes #375

dalepotter committed:
Include legacy organisation test case files in the testrunner

dalepotter committed:
Refactor testrunner to make more DRY

dalepotter committed:
Fix detail in fixture todos

allthatilk commented on tests/test_schemas.py:
```diff
@@ -54,27 +68,42 @@ def load_as_dataset(filepath):
-def test_pass_files(activity_schema, filepath):
-    """Check that all 'should-pass' test files are XML and pass Schema validation."""
+@pytest.mark.parametrize('filepath', list_xml_files_recursively('tests/activity-tests/should-pass/') +  # Legacy activity test cases
+                         list_xml_files_recursively('tests/should-pass/iati-activities/') +  # Legacy activity test cases
```
Is this coment correct? I thought tests/should-pass were the new test cases for 2.03?

dalepotter replied:
Fixed in 684e12b.

allthatilk commented on tests/test_schemas.py:
```diff
@@ -54,27 +68,42 @@ def load_as_dataset(filepath):
 
+    if 'activity-tests' in filepath or 'iati-activities' in filepath:
+        schema = activity_schema
+    else:
```
Not sure the `else` is doing much here, could it be implicit?

dalepotter replied:
Changed to an elif in as explicit is better than implicit in fda3cb9.

allthatilk commented on tests/test_schemas.py:
```diff
@@ -54,27 +68,42 @@ def load_as_dataset(filepath):
 
+    if 'activity-tests' in filepath:
+        schema = activity_schema
+    else:
```
As above

allthatilk replied:
Though might want to extract this as schema detection generally anyway.

dalepotter replied:
Fixed in a86cf85.

allthatilk commented on tests/test_schemas.py:
```diff
@@ -87,7 +116,12 @@ def test_2_03_fail_files(activity_schema, filepath):
     failure_reason = filename_no_extension.split("_")[0]
 
-    error_log = iati.validator.validate_is_iati_xml(dataset, activity_schema)
+    if 'iati-activities' in filepath:
```
I'm guessing this is one of the non-DRY things. Think it's worth extracting it.

dalepotter replied:
Fixed in a86cf85.

allthatilk reviewed (changes requested):
Just a few changes. Think it is worth extracting out schema detection.

dalepotter committed:
Fix comment

Previous comment was incorrect.

dalepotter committed:
Refactor else to be an explicit test for an organisation file

Refactored since an explicit test is better than one that is implicit.

dalepotter committed:
Refactor if/else schema logic into a schema factory

This makes the code more DRY whilst retaining the fact that schemas are only instantinted once (regardless of the number of test cases).

allthatilk commented on tests/should-fail/iati-organisations/iati-organisation/err-not-iati-xml-missing-required-element_no-required-elements-exist.xml:
```diff
@@ -0,0 +1,9 @@
+<iati-organisations version="xx">
+
+  <iati-organisation>
+    <!-- No required elements exist as children of iati-organisation-->
```
Need to add `PASS: ` to start of comment for consistency across 2.03 test data.

dalepotter replied:
Fixed in 303e54f

allthatilk commented on tests/test_schemas.py:
```diff
@@ -54,40 +87,47 @@ def load_as_dataset(filepath):
-def test_pass_files(activity_schema, filepath):
-    """Check that all 'should-pass' test files are XML and pass Schema validation."""
+@pytest.mark.parametrize('filepath', list_xml_files_recursively('tests/activity-tests/should-pass/') +  # Legacy activity test cases
+                         list_xml_files_recursively('tests/should-pass/iati-activities/') +  # Activity test cases in new format
```
new format -> `2.03+ format`

dalepotter replied:
Fixed in 54599ae.

allthatilk commented on tests/test_schemas.py:
```diff
@@ -54,40 +87,47 @@ def load_as_dataset(filepath):
+@pytest.mark.parametrize('filepath', list_xml_files_recursively('tests/activity-tests/should-pass/') +  # Legacy activity test cases
+                         list_xml_files_recursively('tests/should-pass/iati-activities/') +  # Activity test cases in new format
+                         list_xml_files_recursively('tests/organisation-tests/should-pass/') +  # Legacy organisaion test cases
+                         list_xml_files_recursively('tests/should-pass/iati-organisations/'))  # Organisation test cases in new format
```
as above

dalepotter replied:
Fixed in 54599ae.

dalepotter committed:
Make in-code comment clearer

allthatilk commented on tests/test_schemas.py:
```diff
@@ -54,40 +87,47 @@ def load_as_dataset(filepath):
     assert iati.validator.is_xml(dataset)
-    assert iati.validator.is_iati_xml(dataset, activity_schema)
+    assert iati.validator.is_iati_xml(dataset, schema)
+
```
Should this be 2 blank lines? Not sure but it looks odd...

dalepotter replied:
According to the linter, there should be two blank lines between function definitions.

allthatilk replied:
Ah, there is now. Good good!

allthatilk reviewed (changes requested):
Some very small things and then it's good to go!

dalepotter committed:
Add clarity to iati-organisations test case files

allthatilk reviewed (approved):
Looks good!
